aikau/src/main/resources/alfresco/services/FileUploadService.js

```diff
@@ -0,0 +1,708 @@
+         formData.append("overwrite", fileInfo.uploadData.overwrite);
+         formData.append("thumbnails", fileInfo.uploadData.thumbnails);
+         if (fileInfo.uploadData.updateNodeRef) {
+            formData.append("updateNodeRef", fileInfo.uploadData.updateNodeRef);
+         }
+         if (fileInfo.uploadData.description) {
+            formData.append("description", fileInfo.uploadData.description);
+         }
+
+         // Open and send the request
+         fileInfo.request.open("POST", url, true);
+         fileInfo.request.send(formData);
+      },
+
+      /**
+       * Calculates the overall progress of all the uploads and calls the display widget with the data.
+       *
+       * @instance
+       */
+      updateAggregateProgress: function alfresco_services_FileUploadService__updateAggregateProgress() {
+
+         // Setup variables
+         var fileIds = Object.keys(this.fileStore),
+            stillUploading = false,
+            cumulativeProgress = 0;
+
+         // Run through all uploads, calculating total and current progress
+         // NOTE: We don't include failed/completed uploads
+         array.forEach(fileIds, function(fileId) {
+            var fileInfo = this.fileStore[fileId];
+            switch (fileInfo.state) {
+               case this.state.ADDED:
+               case this.state.UPLOADING:
+                  cumulativeProgress += fileInfo.progress;
+                  stillUploading = true;
+                  break;
+               default:
+                  cumulativeProgress += 100;
+            }
+         }, this);
+
+         // Calculate total percentage and send to widget
+         var currentProgress = (cumulativeProgress / this.totalProgressPercent * 100) % 100, // Using modulus to cope with multiple batches
+            currentProgressPercent = stillUploading ? Math.round(currentProgress) : 100;
+         this.uploadDisplayWidget.updateAggregateProgress(currentProgressPercent);
+
+         // If no longer have uploads pending, reset the total progress counter
+         if (!stillUploading) {
+            this.totalProgressPercent = 0;
+         }
+
+         // Update the container title with the aggregate progress if possible
+         var title = this.message(this.uploadsContainerTitle, currentProgressPercent);
+         if (currentProgressPercent === 100) {
+            title = this.message(this.uploadsContainerTitleComplete);
+         }
+         this.uploadsContainerUpdateTopic && this.alfPublish(this.uploadsContainerUpdateTopic, {
+            title: title
+         });
+      },
+
+      /**
+       * Validate the supplied collection of files, sending a notification of any invalid ones, and returning the valid ones.
+       *
+       * @instance
+       * @param {object[]} files The files
+       * @returns {object[]} The valid files
+       */
+      validateFiles: function alfresco_services_FileUploadService__validateFiles(files) {
```

Comment:
Yes, you've explained before why you have things as functions instead of closures.

This is still a function, rather than a closure, however the existing functionality limited one to processing a file individually, rather than looking at all of the files in case one wanted to control what types of files could be uploaded simultaneously. Also, the code now correctly allows a failed-validation message to be put into the display, rather than always saying that it's failed because it's 0kb.

I do not believe this should change back to how it was.